### DISCUSSION TITLE
URL param to allow loading layout from a remote URL

### DISCUSSION THREAD

**Discussion by jtbandes** [foxglove/community#217] [Q&A]
Related: https://github.com/foxglove/community/issues/193

I'd like to embed Studio in my website using an `<iframe>`. I need to be able to specify a layout, but I can't control the localStorage or the user's account, so I'd like to do something like `&layout.data=<base64...>` or `&layout.url=https://example.org/mylayout.json`.

Bonus: API to dynamically change the layout if my website wants to swap between multiple example layouts.

**Comment by defunctzombie** [comment]
Related - loading a set of "team layouts" or "layout templates" from a url? Maybe a separate issue?

**Comment by jhurliman** [comment]
Also related to foxglove/community#194

**Comment by jtbandes** [comment]
Hm, maybe a duplicate, although that other ticket seems to be trying to cover multiple things in one.

**Comment by jameskuszmaul-brt** [comment]
+1 to this. Not sure what the reference to "team layouts" is, but being able to load in with multiple useful layouts available to the user would be great as well. But just getting a single default layout via remote url + deep linking would be amazing.

**Comment by defunctzombie** [comment]
@jameskuszmaul-brt "team layouts" is a reference to the layout sharing feature we have within Studio. When you are signed-in to your Foxglove account in Studio you are able to share layouts with other members of your organization. You can also include the layoutId in urls you share with other members of your org and they will be able to load the same layout.

**Comment by jasonimercer** [comment]
A nice blend between the old webviz feature of 1 hosted layout and the current Foxglove abilities would be to allow us to load a set of named layouts from a URL/json file.

In our webviz use-case we can launch the viewer using different layouts for realtime technical observations of the state of the robots at "dark sites" where there is no internet access. We'd like to transition to Foxglove and the ability to have curated layouts hosted on the robot and switchable at runtime would be very appealing.

**Comment by amacneil** [comment]
Importing comment from @adrienbarral https://github.com/foxglove/studio/issues/2117#issue-1048474869

> In my firm we build ROS1 and ROS2 based robots. We plan to use FoxGlove to propose some maintenance pages for our robots. Most of the time, when users will use these maintenance pages, they will be offline (without any internet connection, and we can't assume that they already had one). I would like to serve foxglove web on a webserver hosted on the robot. And I would like to have some predefined layouts hosted on the robot. These predefined layouts will be designed by the robots development teams. They haven’t vocation to be modifiable.
>
> Today I do this with webviz by using the URL parameter :"layout-url", and having some JSON layout hosted on my robot. That works like a charm. I read the shared gdoc document “FoxGlove studio deeplink”, and my use case doesn't seem to be addressed. (https://docs.google.com/document/d/1jULICvQaZbJM8Et4ylFCE7GIFwedq7YkB4NhvnpjW0k/edit#heading=h.ic2oeiqyrxid)
>
> What I plan to do is to provide a backend service (that we will publish on github (or that can be part of the foxglove repo, I don’t care) ) that will serve some layouts. This service will run on our robots. Then, we could create a new class in FoxGlove Studio that implements the IRemoteLayoutStorage Service to connect to this backend. Of course our backend service will be very simple. We don't plan to address any authentication mechanism, and it will not be designed to be hosted elsewhere than in a local area network. So, someone that needs to use a layout stored on the web and shared with teammates will continue to use your existing service.

**Comment by amffej** [comment]
+1 on the ability to load layout from a file like `&layout.url=https://example.org/mylayout.json`

**Comment by camchambers** [comment]
+1 for loading layouts from a file.

**Comment by maxpolzin** [comment]
Thx for pointing me  to this issue, @jtbandes. +1 for this feature. 

Is there another way to load a layout when offline?
Is there another way to share layouts with non-organization people?

**Comment by adrienbarral** [comment]
@maxpolzin => Did you see that you can export Layout as a JSON file to import them later.

That fullfill a lot of offline usecase.
You can do this by right clicking on your layout, that will open this contextual menu from wich you can export :

![Screenshot_20220706_103754](https://user-images.githubusercontent.com/8557673/177508121-4972138f-8ac3-4315-9664-9f0cebeba021.png)

**Comment by maxpolzin** [comment]
@adrienbarral Thx for the hint, I am aware of this feature. Unfortunately, this requires the layout.json, which I want to import, to be saved on my local machine. That's not the case for me; the layout.json which I want to import is located on a remote machine.

**Comment by markkuskoodi** [comment]
+1 for this feature. We would use the cloud service, but our bag files are only available in our domain which means we need to use self-hosted version of Foxglove. This feature `&layout.url=https://example.org/mylayout.json` would help us a lot in terms of sharing and providing layouts.

**Comment by jhurliman** [comment]
I think this is what was implemented in this commit? https://github.com/foxglove/studio/commit/59abc5e9e6ce74866f9c827e100bfca8e39e7793

That's from a fork of studio. We would need to open a PR against foxglove/studio introducing the feature.

**Comment by siddancha** [comment]
Any insight on how highly the Foxglove team has prioritized this ticket?

IMO loading layouts from a remote URL is the most useful feature that's still missing from Foxglove Studio! I have literally created multiple custom builds of Studio each with a different default layout to get around this limitation 😢. Would +1 more than once if I could!

**Comment by defunctzombie** [comment]
> We would use the cloud service, but our bag files are only available in our domain which means we need to use self-hosted version of Foxglove

@markkuskoodi How does the domain where your bag files are available limit using the hosted version of studio? Also happy to take this to DM on our slack if there's more context outside this ticket.

**Comment by defunctzombie** [comment]
> I have literally created multiple custom builds of Studio each with a different default layout to get around this limitation 😢. Would +1 more than once if I could!

@siddancha Where has the layout storage and sync feature of the foxglove account not met your needs that you had to make custom builds with the layout as part of the build?

**Comment by siddancha** [comment]
@defunctzombie I have two use cases that are not being met:

1. Sharing visualizations publicly with users who are outside my org, or maybe don't even have a Foxglove account. They currently cannot access layouts stored in my account.
2. I want to use different layouts for different types of visualizations. It would be nice if that can happen with a single URL click, instead of the user (if they are in my org that is) needing to find and click on the right layout that I would need to communicate separately.

DM-ing you separately to give an example of my use case.

**Comment by romainreignier** [comment]
Just to mention latest statement about this long standing feature request (almost [1,5 years old](https://github.com/foxglove/studio/issues/2117#issue-1048474869)):
https://github.com/foxglove/studio/pull/4716#issuecomment-1370410644

**Comment by ruffsl** [comment]
Hey folks, I went ahead and took a try at implementing this. Checkout this PR:

- https://github.com/foxglove/studio/pull/5987

I needed this for my own use case as documented and described here:

- https://github.com/ros-planning/navigation2/pull/3551#issuecomment-1541649586

**Reply by RileyEv** [reply]
Unfortunately I tried to get this feature merged a few months ago, which didn't end in success, so its unlikely it'll be merged either :disappointed: but clearly the fact its been implemented 3 times by different people shows its a feature users really do want! Hopefully there's something coming to support this feature soon

**Reply by ruffsl** [reply]
Indeed... When digging into your PR, it looks like several other parties have rolled this feature into their own forks:

- https://github.com/tier4/foxglove-studio/pull/7
- https://github.com/MetaverseIndustries/sviz/pull/2
- https://github.com/collabora/foxglove-studio/commit/59abc5e9e6ce74866f9c827e100bfca8e39e7793

In fact, there seems to be a number of forks popping up to fill this feature gap:

- https://github.com/search?type=code&auto_enroll=true&q=foxglove+%2F%28%3Fi%29%28layoutUrl%7Clayout_url%7Clayout-url%29%2F

It looks like something like this feature existed in foxglove when it was first forked from webviz:

https://github.com/foxglove/studio/blob/a04fd6e2517df9e9ad1ee1395a738d8d82f04b25/app/types/panels.js#L89-L102

Which looks to have originated from this PR, updating packages/webviz-core from an internal repo:

- https://github.com/cruise-automation/webviz/pull/301
  - https://github.com/cruise-automation/webviz/commit/4c76bcba00b44b9d1fdf6e9108d5b7a6f58bed10#diff-720b831eaea4efd665b055a5d4565c7d61475b4c87616948a99356dd84edd922R121-R135
- https://github.com/foxglove/studio/blob/9e90917d6d0ce52c23e7ef293532b8ee1206678b/app/components/PlayerManager.tsx#L247-L264

But was subsequently refactored away later on:

- https://github.com/foxglove/studio/pull/142
  - https://github.com/foxglove/studio/pull/142/files#r1189892180

Let's hope this feature makes an eventual return soon!

**Comment by jtbandes** [comment]
Hi folks, we've just merged a change to allow overriding the default layout when using the Docker image: https://github.com/foxglove/studio/pull/6400. This should be helpful for some of the use cases described above.

**Reply by harderthan** [reply]
@jtbandes 

The link you shared (#6400) seems to be expired or no longer accessible — could you clarify whether this feature has been fully implemented and released?

I’d love to understand the current status and how to use this functionality if it's available. Thanks again!
